compiler/rustc_query_system/src/tls.rs

```diff
@@ -0,0 +1,164 @@
+use crate::dep_graph::TaskDepsRef;
+use crate::query::QueryJobId;
+use rustc_data_structures::sync::{self, Lock};
+use rustc_data_structures::thin_vec::ThinVec;
+use rustc_errors::Diagnostic;
+
+#[cfg(not(parallel_compiler))]
+use std::cell::Cell;
+
+#[cfg(parallel_compiler)]
+use rustc_rayon_core as rayon_core;
+
+/// This is the implicit state of rustc. It contains the current
+/// query. It is updated when
+/// executing a new query.
+#[derive(Clone, Default)]
+struct ImplicitCtxt<'a> {
```

Comment:
You should keep the `tcx` field as a `*const ()`. That lets you move more of the `tls` module intact and the parallel compiler doesn't regress. `with_opt` can convert it back in `rustc_middle`.

You'll want to add an assertion about `GlobalCtxt` being `Sync` too, since the type is now erased from `ImplicitCtxt`.